rfc/0000-configuration-and-shared-files.md

```diff
@@ -0,0 +1,264 @@
+
+- JSON is less human-friendly than TOML and does not support comments which can be useful for documenting configuration files.
+- YAML is more complex than TOML, and can be harder to parse correctly. It is also prone to errors due to the way it handles indentation and somewhat unconventional types (e.g., sexagesimal numbers where `1:2:3` equals `3723`, octal numbers where `0123` equals `83`, and automatic conversion of boolean-like values such as `no`, `on`, `y` into booleans).
+
+Configuration files complement the use of environment variables. They can allow for more complex configuration data to be stored, and can be used to store default values shared across the Pact ecosystem without requiring the end user to set environment variables all the time.
+
+This RFC also proposes to have both local (or per-project) configuration files, and user configuration files. This allows a user of Pact to configure their environment in a way that is consistent across all projects, as well as allowing for project-specific configuration.
+
+For example:
+
+- A user may wish to set the their default Pact Broker to `https://pact-broker.local.company.com` in their user configuration file so that all projects use this value by default.
+- A specific project might still be in development, and use a local Pact Broker at `http://localhost:9292` for testing. This could be set in the project's local configuration file.
+- The CI/CD pipeline still configures the Pact Broker using environment variables, as it is not practical to store configuration files in the pipeline.
+
+#### Precedence
+
+The following order of precedence _must_ be followed when reading configuration data:
+
+1. Environment variables
+2. Configuration files, in the following order:
+   1. If an explicit configuration file is set, only that file should be read.
+   2. If no explicit configuration file is set, the following files should be read in order:
+      1. Local configuration file
+      2. User configuration file
+3. Default values
+
+The explicit configuration file may either be set by the `PACT_CONFIG_FILE` environment variable, or by a command line option.
+
+#### Location
+
+To ensure a consistent experience across the Pact ecosystem, the following configuration files are proposed to be used across the Pact ecosystem:
+
+- **User Configuration File**
+
+  Location:
+
+  - Unix and Unix-like systems:
+    - `$XDG_CONFIG_HOME/pact/config.toml` (which defaults to `~/.config/pact/config.toml` if `XDG_CONFIG_HOME` is not set)
+  - Windows:
+    - `%APPDATA%\pact\config.toml`
+  - Fallback:
+    - `~/.pact/config.toml`
+
+  > [!NOTE]
+  >
+  > macOS is considered a Unix-like system for the purposes of this RFC. Should an end-user prefer to use a different location (such as `~/Library/Preferences`), they may do so by setting the `XDG_CONFIG_HOME` environment variable.
```

Comment:
An end-user could set either option, though they would be slightly different. The following would be equivalent:

- `XDG_CONFIG_HOME=$HOME/Library/Preferences` as the result Pact config directory becomes `~/Library/Preferences/pact` and the config file would be `~/Library/Preferences/pact/config.toml`
- `PACT_CONFIG_FILE=$HOME/Library/Preferences/pact/config.toml`

The `PACT_DATA_HOME` is not relevant when parsing config files. Though it might be worth amending the RFC such that the fallback should be `$PACT_DATA_HOME/config.toml` if `PACT_DATA_HOME` is set, and finally to fallback to `~/.pact/config.toml`.

The reason `PACT_DATA_HOME` is typically not used is that Linux and Unix-based systems separate the config from data files (e.g., you would never have a config file in `/usr/lib` or `/usr/share`, and it is always found in `/etc`). The same convention is in effect for user config files, with the files stored in `~/.config` and data files in `~/.local`.

Reply:
>The reason PACT_DATA_HOME is typically not used is that Linux and Unix-based systems separate the config from data files (e.g., you would never have a config file in /usr/lib or /usr/share, and it is always found in /etc). The same convention is in effect for user config files, with the files stored in ~/.config and data files in ~/.local.

Okay this makes sense, did a bit of background reading on macos/XDG_CONFIG_HOME with regards to ported unix tools and expected behaviour and your reasoning seems sound. I think the general expectation is someone can backup their config files without having to get a load of app data alongside it?